Roll sync async engine

floitschG opened the pull request.

floitschG committed:
Prepare everything for sync-async.

floitschG committed:
Prepare everything for sync-async.

floitschG committed:
Roll the engine.

This engine includes a VM that defaults to sync-async.

flutter/engine@c8eeee4 Roll the Dart VM.

mravn-google commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -507,6 +507,12 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
+    // Delay this function by a microtask.
+    // Otherwise it will open as cope immediately, which then open the
+    // `asyncBarrier` is invoked. The `asyncBarrier` is immediately following
+    // the call to `testZone.runBinary(_runTestBody`, so delaying by one
```
This is an older version of your comment, with a missing right parenthesis and unparseable text at the beginning

floitschG replied:
Thanks.
I somehow got my versions mixed up...

mravn-google reviewed (approved):
LGTM

floitschG committed:
Merge branch 'fix_tests_for_sync_async' of github.com:floitschG/flutter into roll_sync_async_engine

floitschG committed:
Fix comment.

floitschG committed:
Trigger

floitschG committed:
Trigger

floitschG committed:
Trigger

Hixie commented on dev/automated_tests/flutter_test/test_async_utils_guarded_test.dart:
```diff
@@ -14,7 +14,10 @@ class TestTestBinding extends AutomatedTestWidgetsFlutterBinding {
+    await tester.pumpWidget(const Directionality(
+      textDirection: TextDirection.ltr,
+      child: const Text('Hello'),
+    ));
```
we should just pass `textDirection: TextDirection.ltr` to the `Text` widget rather than adding `Directionality`.

Hixie commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -507,6 +507,12 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
+    // the `asyncBarrier` is invoked. The `asyncBarrier` is immediately
+    // following the call to `testZone.runBinary(_runTestBody)`, so delaying
+    // by one microtask is enough to ensure that the timing is correct.
+    await new Future<Null>.microtask(() {});
```
I don't really understand this comment. Can you extend it to more fully explain the sequence of events?

Hixie replied:
I think what I'm having trouble with is the question of what the "open scope" that this comment is referring to is. Where do we open a scope which delaying the code here helps with? This microtask doesn't interact with the async barrier, since the async barrier isn't called until later within this method (synchronously) anyway. Indeed, the async barrier can't possibly be what drives the microtasks (as per the last sentence of the comment here) since it isn't invoked until after the microtasks are invoked, since it's called synchronously by this method.

Hixie commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -848,6 +854,13 @@ class AutomatedTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
 
     return new Future<Null>.microtask(() async {
+      // Run all queued microtasks.
+      await new Future<Null>.microtask(() {});
```
I don't understand the purpose of this initial delay. This whole function is already run in a microtask.

Hixie replied:
But all the microtasks in _runTest are run in the FakeAsync zone, no? What actual microtasks get scheduled in this zone? The whole point of this loop is to flush the microtasks, which we do with an await in the while loop; why does pushing us to the end of the event loop help here?

floitschG replied:
Ok. I looked at this in more detail.

So the comment in line 875 was wrong.
It stated that by returning the `testBodyResult` future from a future that was created outside the `fakeAsync` it would not need the `fakeAsync` anymore.
That's not true. The `testBodyResult` future might still get a `.then` call. It just happened not to be the case, because we have an optimization that propagates results without explicitly invoking `.then` if we know already the next listener.

By delaying twice the next listener was already known. By delaying only once, an internal `.then` call was added to the `testBodyResult`. However, since there was no additional `fakeAsync.fushMicrotask` call that `.then` call was never executed.

I refactored the code.

Hixie replied:
ah, this makes sense.

Hixie commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -848,6 +854,13 @@ class AutomatedTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
+      // When the test had an exception, the test-framework already ran the
+      // teardown functions, removing the _fakeAsync object.
+      if (fakeAsync == null)
+        return null;
```
Why is this a problem now but not before?

floitschG replied:
Seems like the code around it has changed since I wrote this patch. That part is removed now.

Hixie reviewed (changes requested):
see comments above

floitschG committed:
Address comments.

floitschG committed:
Pass `TextDirection` directly to `Text`.

floitschG committed:
Fix bad timing dependencies.

floitschG committed:
Fix for lints.

Hixie commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -507,6 +507,17 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
+    // By delaying by one microtask, the `asyncBarrier` will not yet see the
+    // open scope and also ensure that the all microtasks (including this one)
+    // are flushed.
+    await new Future<Null>.microtask(() {});
```
(github ate my comment so I'm going to copy and paste it here, my apologies:)

I think what I'm having trouble with is the question of what the "open scope" that this comment is referring to is. Where do we open a scope which delaying the code here helps with? This microtask doesn't interact with the async barrier, since the async barrier isn't called until later within this method (synchronously) anyway. Indeed, the async barrier can't possibly be what drives the microtasks (as per the last sentence of the comment here) since it isn't invoked until after the microtasks are invoked, since it's called synchronously by this method.

floitschG replied:
The `_runTestBody` function calls `runApp` followed by `pump` which seems to open a scope.
After the call to `pump` it yields (since it's doing an `await`), and the controlflow continues in `_runTest` (line 504 above) where it does a `timeout?.catchError` followed by an `asyncBarrier`.
The `asyncBarrier` checks that no scope is open, which is not the case, since `pump` just opened one.

By delaying the `_runTestBody` function the `_runTest` continues *first* with the `timeout?.catchError` and `asyncBarrier` (now succeeding) before running the `_runTestBody`.

floitschG replied:
This shouldn't be an issue anymore.

Hixie replied:
ah, interesting. sounds like the old code was just wrong too, then. The purpose of the guards is to catch open scopes in the test that weren't fully awaited.

Hixie commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -507,6 +507,17 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
+    // Delay this function by a microtask.
+    //
+    // The `_runTest` functions invokes `_runTestBody` and then invokes
+    // `asyncBarrier()` afterwardes. The `asyncBarrier` verifies that no scope is
```
nit: `afterwards`

floitschG replied:
done.

Hixie commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -848,22 +859,23 @@ class AutomatedTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
-      // by returning the wrapped microtask future that we've created _outside_
-      // the fake async zone.
-      return testBodyResult;
+      return resultFuture;
```
yeah this is much better, thanks.

floitschG committed:
Fix typo.

floitschG committed:
Reorder code.

This avoids lengthy comments and an `await new Future.microtask`.
Thanks Slava.

Hixie commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -498,11 +498,11 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
     timeout?.catchError(handleUncaughtError);
-    asyncBarrier(); // When using AutomatedTestWidgetsFlutterBinding, this flushes the microtasks.
     return testCompleter.future;
   }
```
I don't understand this change. What does it even mean to have an async barrier before the test runs? The whole point here is to catch the case of people having guarded calls that are still open after the test has returned.

Maybe the line should be removed and moved to the end of runTest?

floitschG replied:
done.

floitschG committed:
Move asyncBarrier.

mravn-google commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -848,22 +848,23 @@ class AutomatedTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
+      // of the fake-zone below, the new future will be completed and can then
+      // be used without fakeAsync.
+      final Future<Null> resultFuture = testBodyResult.then<Null>((_) {
+        /* Do nothing. */
```
[nit] Any reason this comment isn't just `// Do nothing.`?

floitschG replied:
No reason.
Changed it.

floitschG committed:
Trigger

floitschG committed:
Change comment style.

Hixie commented on packages/flutter_test/lib/src/binding.dart:
```diff
@@ -532,6 +531,7 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
     }
 
     assert(inTest);
+    asyncBarrier(); // When using AutomatedTestWidgetsFlutterBinding, this flushes the microtasks.
```
looks like we actually already have one here, so it's not needed after all (see line 519)

Hixie commented:
Ideally we'd add a test to verify that the async barrier at the end is working (e.g. verify that a test that just does an unawaited pump actually triggers the warning), but we can do that in another PR.

floitschG committed:
Don't cache pub.

floitschG committed:
Undo changes to travis file.
